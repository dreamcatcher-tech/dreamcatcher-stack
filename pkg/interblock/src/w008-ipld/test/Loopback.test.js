import { assert } from 'chai/index.mjs'
import { Loopback, Request, Address, Channel, Reply, Network } from '..'
import Debug from 'debug'
const debug = Debug('interblock:tests:Channel')
Debug.enable()

describe('loopback', () => {
  test('basic', async () => {
    let loopback = Loopback.create()
    loopback = await loopback.crush()
    const ping = Request.create('PING')
    loopback = loopback.txRequest(ping)
    const channelId = Network.FIXED_IDS.LOOPBACK
    assert.strictEqual(channelId, loopback.channelId)
    const rxRequest = loopback.rxReducerRequest()
    assert.strictEqual(rxRequest.requestIndex, 0)
    assert.strictEqual(rxRequest.channelId, channelId)
    assert.strictEqual(rxRequest.stream, 'reducer')
    assert.strictEqual(rxRequest.type, ping.type)

    const pong = Reply.create()
    loopback = loopback.txReducerReply(pong)
    assert(!loopback.rxReducerRequest())
    const rxReply = loopback.rxReducerReply()
    assert.strictEqual(rxReply.type, pong.type)

    loopback = loopback.shiftReducerReply()
    assert(!loopback.rxReducerReply(channelId))
    const { reducer: rxReducer } = loopback.rx
    assert.strictEqual(rxReducer.requestsLength, 1)
    assert.strictEqual(rxReducer.repliesLength, 1)
    const { reducer: txReducer } = loopback.tx
    assert.strictEqual(txReducer.requestsLength, 0)
    assert.strictEqual(txReducer.repliesLength, 0)

    const last = await loopback.crush()
    expect(last).toMatchSnapshot()
  })
})
