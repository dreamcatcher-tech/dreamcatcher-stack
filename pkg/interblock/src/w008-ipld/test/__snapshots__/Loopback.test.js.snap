// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loopback basic 1`] = `
Loopback {
  "address": Address {},
  "aliases": Array [],
  "channelId": 0,
  "rx": Rx {
    "reducer": RxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 1,
      "requests": Array [],
      "requestsLength": 1,
    },
    "system": RxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 0,
      "requests": Array [],
      "requestsLength": 0,
    },
  },
  "tx": Tx {
    "reducer": TxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 0,
      "requests": Array [],
      "requestsLength": 0,
    },
    "system": TxQueue {
      "promisedReplies": Array [],
      "promisedRequestIds": Array [],
      "replies": Array [],
      "repliesLength": 0,
      "requests": Array [],
      "requestsLength": 0,
    },
  },
}
`;
